Use default export settings when creating new exports if they exist

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Default Export Settings](https://dimagi-dev.atlassian.net/browse/SAAS-11474)

The creation of a `DefaultExportSettings` object is behind a feature flag. If a valid settings object is returned from the `get_or_create_default_export_settings_for_domain` method, the new path is executed which includes overriding hard coded default export settings with settings specified in the account's `DefaultExportSettings` object.

I tested this on a local branch that contains commits from both here and [the UI portion of this feature](https://github.com/dimagi/commcare-hq/pull/28941). 
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
DEFAULT_EXPORT_SETTINGS
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
The changes that would result in a different code flow are behind a feature flag, so users should not experience any difference once this is rolled out. 
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
if default settings exist use them when creating new exports

gherceg commented on corehq/apps/export/dbaccessors.py:
```diff
@@ -181,6 +181,19 @@ def _properly_wrap_export_instance(doc):
 
+def get_default_export_settings_for_domain(domain):
+    from .models import DefaultExportSettings
+    from corehq.apps.accounting.models import Subscription, SoftwarePlanEdition
```
I think this is the appropriate place for this db helper, but not sure if it is bad practice to import models from another app.

biyeun replied:
I would call it `get_or_create_default_export_settings_for_domain`. Also I would lean toward this being a utility vs a dbaccessor as the dbaccessors are really meant for interfacing with couch <> SQL backends

gherceg commented on corehq/apps/export/models/export_settings.py:
```diff
@@ -12,6 +14,19 @@ class ExportFileType(object):
         (EXCEL_PRE_2007, ugettext_lazy("Excel (older versions)")),
     )
 
+    @classmethod
+    def map_to_couch_type(cls, filetype):
+        from couchexport.models import Format
+
+        if filetype == ExportFileType.EXCEL_2007_PLUS:
+            return Format.XLS_2007
+        elif filetype == ExportFileType.EXCEL_PRE_2007:
+            return Format.XLS
+        elif filetype == ExportFileType.CSV:
+            return Format.CSV
+        else:
+            raise ValueError(f"{filetype} is not supported for export")
+
```
Not thrilled about this and I should have had better foresight when creating `DefaultExportSettings`. I see a couple of paths foward:
- the one implemented above - map these types to format strings used in couch
- update/migrate `DefaultExportSettings` to directly store these values instead of the `ExportFileType` choices I created

biyeun replied:
I would call this method `get_file_type`

orangejenny replied:
I'd lean towards doing the migration, since it sounds pretty straightforward and would eliminate this code.

gherceg replied:
I also prefer doing the migration now and removing the mapping. Just to be clear about what this entails, this file in the `export` app would need to reference code in ex-`submodules/couchexport` [here](https://github.com/dimagi/commcare-hq/blob/579b52f1037d30c8e9184214f7278bb1d6a8b9d2/corehq/ex-submodules/couchexport/models.py#L8-L17). I'm assuming this is fine but just want to make sure. 

@biyeun are you opposed to the migration at all?

orangejenny replied:
I think that import is fine, especially given that [this one](./corehq/apps/export/models/new.py) exists.

biyeun replied:
migration sounds like a good plan

gherceg replied:
Created [this PR](https://github.com/dimagi/commcare-hq/pull/28982) to do the migration.

gherceg commented on corehq/apps/export/views/utils.py:
```diff
@@ -215,8 +216,12 @@ def terminology(self):
+        settings = get_default_export_settings_for_domain(schema.domain)
         instance.is_odata_config = True
         instance.transform_dates = False
+        if hasattr(instance, 'include_errors'):
+            instance.include_errors = settings.odata_include_duplicates
+        instance.split_multiselects = settings.odata_expand_checkbox
```
I don't particularly like checking if the class has that attribute, but only one of the `ExportInstance` subclasses has this attribute so I decided to seize the opportunity to write this logic once instead of in the 2 subclasses.

Though looking at this again, it might be better to add this logic directly to the subclasses here:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/export/views/new.py#L314-L333

biyeun replied:
agreed

biyeun commented on corehq/apps/export/models/new.py:
```diff
@@ -88,8 +87,9 @@
     get_form_inferred_schema,
     get_latest_case_export_schema,
-    get_latest_form_export_schema,
+    get_latest_form_export_schema, get_default_export_settings_for_domain,
```
new line

gherceg committed:
peer review edits

- only access default export settings for odata form exports
- minor method name fixes
- remove unused imports

gherceg committed:
add documentation

gherceg committed:
put creation of default export settings behind feature flag

orangejenny commented on corehq/apps/export/models/new.py:
```diff
@@ -1120,10 +1123,19 @@ def identifier(self):
-            domain=schema.domain,
-            case_type=schema.case_type,
-        )
+        settings = get_or_create_default_export_settings_for_domain(schema.domain)
```
Look like you and @biyeun already discussed this naming, but my take is that seeing `get_or_create` makes me surprised to then see `None` be returned, since the similarly-named django ORM method always return something. `get_or_create_default_export_settings_for_domain_unless_not_allowed` would be a bit much. What about `get_default_export_settings_if_available` or even just `get_default_export_settings`?

gherceg replied:
I went with `get_default_export_settings_for_domain`. Thought it might be nice to be more explicit and start with `try_to_get...` or something similar but I think this is sufficient. Thanks for the feedback.

orangejenny reviewed (dismissed):
Didn't mean to click approve, just because Biyeun knows this code better. I don't have specific changes to request, I just don't know of another way to remove the approval.

orangejenny commented on corehq/apps/export/utils.py:
```diff
@@ -51,3 +52,22 @@ def get_export(export_type, domain, export_id=None, username=None):
     raise Exception("Unexpected export type received %s" % export_type)
+
+
+def get_or_create_default_export_settings_for_domain(domain):
```
I do think this is a fine django app for this function to live in. Export seems more logical than accounting. The other place I could see it living is the new enterprise app, but export is also fine.

gherceg committed:
test default export settings

gherceg committed:
properly setup tests

gherceg commented:
Do not merge until [this migration](https://github.com/dimagi/commcare-hq/pull/28982) is merged.

gherceg committed:
finally fixed tests

gherceg committed:
remove unused import

gherceg committed:
rename helper method for obtaining default export settings

gherceg committed:
Merge branch 'master' into gh/use-default-export-settings

gherceg committed:
remove get_file_format

biyeun reviewed (approved):
looks great!

gherceg committed:
update mock patch method name
